From a9d15396634c4bc88e8b7645ed7e4baf23de2eda Mon Sep 17 00:00:00 2001
From: Joe Zhou <shjzhou@marvell.com>
Date: Tue, 22 Apr 2014 10:04:11 +0800
Subject: [PATCH 0662/1240] fix: usb: fix USB3.0 connect-disconnect detection
 issue

1. After U-Boot is up, plug out and plug in USB3 device,
   it cannot be detected by "usb start" command
   The PORTSRC register value is 0x281203, BIT 17 CSC is 0
   which indicates "no connection state change".
2. If plug out and plug in USB3 device during "scanning"
   after input "usb start" command, it can be detected.
   the PORTSRC register value is 0x2a1203, BIT 17 CSC is 1
   which indicates "Connection" and "Connection state change".

Root cause:
CSC bit is not set if USB3 device is connected before usb
reset operation. "usb start" command does reset operation,
so CSC bit is not set in case 1.

Solution:
CSC bit should be ignored when CCS is set in function
usb_hub_configure.

JIRA: SYSTEMSW-443 & SYSTEMSW-513

Change-Id: I176493e80157c876e9272775a84aaa2c93e8905f
Signed-off-by: Joe Zhou <shjzhou@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/7150
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: Omri Itach <omrii@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/26700
Reviewed-by: Kostya Porotchkin <kostap@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
---
 common/usb_hub.c | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/common/usb_hub.c b/common/usb_hub.c
index 66b4a72..ebf2380 100644
--- a/common/usb_hub.c
+++ b/common/usb_hub.c
@@ -429,9 +429,19 @@ static int usb_hub_configure(struct usb_device *dev)
 
 			portstatus = le16_to_cpu(portsts->wPortStatus);
 			portchange = le16_to_cpu(portsts->wPortChange);
-
+#ifdef CONFIG_MVEBU
+/* If usb device was already connected before detection ('usb start'),
+ * port connect status bit is set (USB_PORT_STAT_CONNECTION),
+ * but port connect status change bit (USB_PORT_STAT_C_CONNECTION) is not set.
+ * - Added WA (aligned with kernel driver): instead of checking port change status,
+ *   to be satisfied with checking USB_PORT_STAT_CONNECTION Bit on portstatus only" */
+			if (((portchange & USB_PORT_STAT_C_CONNECTION) ==
+				(portstatus & USB_PORT_STAT_CONNECTION)) ||
+				(portstatus & USB_PORT_STAT_CONNECTION))
+#else
 			if ((portchange & USB_PORT_STAT_C_CONNECTION) ==
 				(portstatus & USB_PORT_STAT_CONNECTION))
+#endif
 				break;
 
 		} while (get_timer(start) < CONFIG_SYS_HZ * 10);
@@ -442,7 +452,17 @@ static int usb_hub_configure(struct usb_device *dev)
 		debug("Port %d Status %X Change %X\n",
 		      i + 1, portstatus, portchange);
 
+#ifdef CONFIG_MVEBU
+/* If usb device was already connected before detection ('usb start'),
+ * port connect status bit is set (USB_PORT_STAT_CONNECTION),
+ * but port connect status change bit (USB_PORT_STAT_C_CONNECTION) is not set.
+ * - Added WA (aligned with kernel driver): instead of checking port change status,
+ *   to be satisfied with checking USB_PORT_STAT_CONNECTION Bit on portstatus only" */
+		if ((portchange & USB_PORT_STAT_C_CONNECTION) ||
+			(portstatus & USB_PORT_STAT_CONNECTION)) {
+#else
 		if (portchange & USB_PORT_STAT_C_CONNECTION) {
+#endif
 			debug("port %d connection change\n", i + 1);
 			usb_hub_port_connect_change(dev, i);
 		}
-- 
1.9.1

